overlord/clusterstate/clusterstate.go

```diff
@@ -23,26 +23,215 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"io"
+	"strconv"
 
 	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/overlord/assertstate"
 	"github.com/snapcore/snapd/overlord/devicestate"
 	"github.com/snapcore/snapd/overlord/snapstate"
 	"github.com/snapcore/snapd/overlord/state"
 )
 
 var (
-	installWithGoal   = snapstate.InstallWithGoal
-	removeMany        = snapstate.RemoveMany
-	updateWithGoal    = snapstate.UpdateWithGoal
-	storeUpdateGoal   = snapstate.StoreUpdateGoal
-	storeInstallGoal  = snapstate.StoreInstallGoal
-	devicestateSerial = devicestate.Serial
+	installWithGoal  = snapstate.InstallWithGoal
+	removeMany       = snapstate.RemoveMany
+	updateWithGoal   = snapstate.UpdateWithGoal
+	storeUpdateGoal  = snapstate.StoreUpdateGoal
+	storeInstallGoal = snapstate.StoreInstallGoal
 )
 
-// ApplyClusterState creates the tasks needed to apply the state described by
+// ErrNoClusterAssertion indicates there is no current cluster assertion available.
+var ErrNoClusterAssertion = errors.New("clusterstate: no cluster assertion")
+
+// clusterState contains the state of clustering on this device.
+//
+// TODO: This is pretty verbose right now, but we might want to put everything
+// under one namespace in the state?
+type clusterState struct {
+	// Current contains the information needed to find the current cluster
+	// assertion. Maybe we should consider some sort of sequence container, like
+	// we use in snapstate?
+	Current clusterAssertionState `json:"current"`
+}
+
+// clusterAssertionState contains the information needed to find a specific
+// cluster assertion.
+type clusterAssertionState struct {
+	// ClusterID is the globally unique identifier for this cluster assertion.
+	ClusterID string `json:"cluster-id"`
+	// Sequence is the sequence point at which the cluster assertion with the
+	// ClusterID.
+	Sequence int `json:"sequence"`
+	// AuthorityID is the ID of the account that is associated with this
+	// cluster. When updating the cluster to a new sequence number, the new
+	// cluster assertion must be signed by an account-key associated with the
+	// same account.
+	AuthorityID string `json:"authority-id"`
+}
+
+// InitializeNewCluster installs a new cluster assertion bundle, replacing any
+// existing state. Callers must hold the state lock.
+func InitializeNewCluster(st *state.State, bundle io.Reader) error {
+	batch, cluster, err := decodeClusterBundle(bundle)
+	if err != nil {
+		return err
+	}
+
+	var existing clusterState
+	if err := st.Get("cluster", &existing); err != nil && !errors.Is(err, state.ErrNoState) {
```

Comment:
sorry, I probably confused myself and you,  what's the agreement on the spec that a device can only belong to one cluster in which case this is always an error if current is already set?